[FEATURE] Add SuiteParameterDict to all Expectation Kwargs

Pascal06S opened the pull request:
Adds SuiteParameterDict to the type definition of all expectation arguments. By doing that, all kwargs of an expectation are parametrizable. See discussion: https://app.slack.com/client/T7NEH00BA/CUTCNHN82

- [ ] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [ ] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB], [MINORBUMP]
- [ ] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [ ] Appropriate tests and docs have been updated

For more information about contributing, visit our [community resources](https://docs.greatexpectations.io/docs/core/introduction/community_resources#contribute-code-or-documentation).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

Pascal06S committed:
add suite parameter dict to all expectation kwargs

Pascal06S committed:
remove unnecessary files

Pascal06S committed:
Merge branch 'great-expectations:develop' into feature/suite_parameter_dict

Pascal06S committed:
linting

Pascal06S committed:
Merge branch 'feature/suite_parameter_dict' of https://github.com/Pascal06S/great_expectations into feature/suite_parameter_dict

joshua-stauffer commented on tasks.py:
```diff
@@ -125,7 +125,7 @@ def lint(
     output_format: Literal["full", "concise", "github"] | None = None,
     watch: bool = False,
-    pty: bool = True,
+    pty: bool = False,
```
this should be reverted - it allows colorized output of the ruff command to the terminal. If for some reason this is causing problems for you locally, you could try ` invoke lint --no-pty`.

joshua-stauffer reviewed (changes requested):
Thanks for the PR, @Pascal06S ! 

Other than the change to the invoke flag, the code look reasonable, but we'll need test coverage to merge this. Most of these code paths are getting dynamic parameters for the first time, so we need to ensure the rest of the system accounts for them.

I would suggest adding a test case for the updated Expectations in `tests/integration/data_sources_and_expectations/expectations`. Each Expectation should already have a module there.

Pascal06S committed:
change back pty, add suite params to QuantileRange

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Pascal06S commented:
> Thanks for the PR, @Pascal06S !
> 
> Other than the change to the invoke flag, the code look reasonable, but we'll need test coverage to merge this. Most of these code paths are getting dynamic parameters for the first time, so we need to ensure the rest of the system accounts for them.
> 
> I would suggest adding a test case for the updated Expectations in `tests/integration/data_sources_and_expectations/expectations`. Each Expectation should already have a module there.

The invoke flag part was mistakenly left in, thanks for pointing it out!

Regarding the test coverage, I am not sure what kind of tests would be best to do for that feature. I went through the mentionend folder to look how params are tested that already had a SuiteParameterDict  in their type definition, but the only example i could find was in https://github.com/great-expectations/great_expectations/blob/ffc424a8a5423bf40788da2d03e7e2fd8e720327/tests/integration/data_sources_and_expectations/expectations/test_expect_table_row_count_to_be_between.py#L134 which did not help me a lot.

What kind of success criteria would be required in the new tests?

joshua-stauffer commented:
The minimum test coverage would just ensure that the expectation validates as expected and GX doesn't blow up when a suite parameter is provided. Here's a quick example PR i threw together that tests both success and failure results, but just testing success the success path is probably sufficient: https://github.com/great-expectations/great_expectations/pull/11225

Pascal06S committed:
Merge branch 'great-expectations:develop' into feature/suite_parameter_dict

Pascal06S committed:
add pytests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

joshua-stauffer commented on tests/integration/data_sources_and_expectations/expectations/test_expect_query_results_to_match_comparison.py:
```diff
@@ -948,3 +948,61 @@ def _create_table_rendered_atomic_content(
+
+    result = multi_source_batch.validate(
+        expectation, expectation_parameters={suite_param_key: suite_param_value}
+    )
```
The mypy error here and below on line 1005 is correct. You'll need to update them to be `multi_source_batch.base_batch.validate(...`

joshua-stauffer commented:
hey @Pascal06S , thanks for adding the tests! I'll give this a thorough review once tests are passing, looks like mypy caught a bug in two of the test implementations.

Pascal06S committed:
fix prescriptive_renderer for mostly and quantiles, fix pytests

Pascal06S committed:
Merge branch 'feature/suite_parameter_dict' of https://github.com/Pascal06S/great_expectations into feature/suite_parameter_dict

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Pascal06S commented:
Thanks @joshua-stauffer i updated the tests! I also fixed a bug in the prescriptive renderer for the already parametrizable params "mostly" and "quantiles"

joshua-stauffer commented:
ok, great, we're getting closer. Since this PR updates the public Expectation API, we need to regen their schemas. Looks like that's the main cause of test failures. To fix, run `invoke schemas --sync` and commit the result.

joshua-stauffer commented:
I see that there are also some failures with the multi source expectation tests. They're occurring because the arbitrary table name you're passing in is not a valid table. To access the secondary table name, use `multi_source_batch.comparison_table_name`. Probably best to just remove the parametrized fixture, since you won't have the table name in scope there.

Pascal06S committed:
update schemas, fix tests

Pascal06S committed:
Merge branch 'feature/suite_parameter_dict' of https://github.com/Pascal06S/great_expectations into feature/suite_parameter_dict

Pascal06S committed:
fix pytests hopefully final

Pascal06S committed:
fix UnexpectedRowsExpectation

Pascal06S committed:
fix intergrity workflow test

Pascal06S commented:
I hope the tests are successful now. They did not fail on my local python 3.10 env. After installing python 3.9 they failed and i think i fixed it now

tyler-hoffman reviewed (approved):
Thanks for doing this! It's been something I've wanted to get to for a while now!

Looks like a lot of tests are unnecessarily parameterized given that there's only one test case in the call to `parametrize`, but I think that's fine; it gives us a spot if we want to add more parameterized tests down the road.

tyler-hoffman committed:
Merge branch 'develop' into feature/suite_parameter_dict

joshua-stauffer reviewed (approved):
looks great @Pascal06S - thanks for the contribution, and for addressing all the feedback. much appreciated!

tyler-hoffman committed:
Merge branch 'develop' into feature/suite_parameter_dict
